### ISSUE TITLE
knex.raw escaping very wrong

### ISSUE THREAD

**Issue by andrecavallari** [tgriesser/knex#2567]
# Bug

I am trying to use knex.raw on postgresql database to create a concatenated search on database, but knex.raw escapes incorrectly because it escapes every dot

## Error message

SQL error message for syntax error

## What I get and what is expected

knex.raw('?? ILIKE ALL (ARRAY[??])',['CONCAT(cars.model,cars.version,cars.year)',['%focus%','%hatch%']])
Produces:  ...  where "CONCAT(cars"."model,cars"."version,cars"."year)" ILIKE ALL (ARRAY["%focus%", "%hatch%"]) ...
Should be: ... where CONCAT("cars"."model","cars"."version","cars"."year") ILIKE ALL (ARRAY['%focus%', '%hatch%']) ...

Also I tried to change ?? to ?, no success too since it escapes as value

I Think, since it's a raw query, should not escape anything

**Comment by elhigu** [comment]
Not a bug. Works just as expected. You are passing column name `CONCAT(cars.model,cars.version,cars.year)` there.

Please read carefully the docs how to use raw and its value `?` and identifier `??` bindings http://knexjs.org/#Raw-Bindings.

Further questions how to use knex should go to stackoverflow. Github issues are for development new features and bug reports etc.

**Comment by andrecavallari** [comment]
Ok buddy, so since it's not a bug, please help me to fix it...

let fields = ['cars.model','cars.version','cars.year'];
let params = ['%ford%','%focus%'];

knex.raw('CONCAT(??) ILIKE ALL (ARRAY[?])',[fields,params])
converts to ARRAY['%ford%,%focus%']

knex.raw('CONCAT(??) ILIKE ALL (ARRAY[?])',[fields,params])
converts to ARRAY["%ford%,%focus%"]

The expected:
ARRAY['%ford%','%focus%"]

Note: The real params is built by request.params from express:
request.query.q.split(' ').map(query => {return `%${query}%`}).join(',')

**Comment by kibertoad** [comment]
@andrecavallari It is generally a good idea to maintain professional tone of conversation in open-source community.
As was already mentioned, GitHub is not intended for providing assistance to consumers of a library, it is a development tool. Please use stackoverflow/forums/etc to get help.

**Comment by elhigu** [comment]
In stackoverflow there are more people answering knex usage questions than here.

**Comment by israelidanny** [comment]
@elhigu I think you're missing the point though. Not *every* single time a question mark is introduced, it should be replaced and turned into a binding. I'm running into the same exact issue with trying to migrate some 'CREATE FUNCTION' statements, and my functions contain some regex and stuff, where question marks are part of the syntax. 

Surely there has to be a way to disable this behavior? There are plenty of use cases where one would explicitly would not want raw to replace ?'s with bindings.

**Comment by kibertoad** [comment]
@IsaacLessard You mean `To prevent replacement of ? one can use the escape sequence \\?.`, right? @elhigu

**Comment by israelidanny** [comment]
@elhigu @kibertoad well, technically you're right. In my case that's what I'm doing - I'm loading up an SQL source file, and since that file should stay a manual file - I end up doing a string replace on it.

It feels like a bit of a hack though.

Don't you think that allowing an unescaped query is appropriate for use cases when one just wants to run an SQL script?

**Comment by israelidanny** [comment]
I see what you mean.

The behavior you described seems optimal (no bindings passed - no
replacement)

Can I submit a failing test for that?

**Comment by kibertoad** [comment]
@israelidanny Any update on the failing test part? I would like to start working on this in the nearby future, test would make sure that I would be solving the right problem :)

**Comment by gambhirpulkit** [comment]
I pass jsonb in knex.raw. It is replacing all question marks with $variable. It is difficult and expensive to escape `?` in 100 keys of postgres json values

**Comment by kibertoad** [comment]
@gambhirpulkit PRs for this functionality or at least failing tests would speed up the process immensely.

**Comment by gambhirpulkit** [comment]
@kibertoad I added this `jsonString.replace(/?/g,'\\?')}` to `jsonString` and it escapped all `?`.

**Comment by kibertoad** [comment]
And this is not breaking the intended part of variable replacement logic? Do all tests pass?

**Comment by gambhirpulkit** [comment]
Yes. I do not understand what you mean exactly by failing tests. Can you please explain that term?
Below is the query that worked now. It was failing without string `replace`
```
bookshelf.knex.raw(`UPDATE laundry_store SET message = jsonb_set(message, '{${req.query.lang}}', '${JSON.stringify(req.body).replace(/?/g,'\\?')}', false) where id=${store.toJSON().id}`)
              .then(res => resolve(res))
              .catch(err => reject(err));
```

**Comment by kibertoad** [comment]
@gambhirpulkit Ah, you mean you did the change in your code, not in knex. I was wondering if automated tests of knex (https://en.wikipedia.org/wiki/Unit_testing) are failing after change in knex, but there were no changes.
Unfortunately, we can't just do the replace within knex in all cases, as that variable replacement is very much intentional, but yes, we need to handle case with no bindings better.

**Comment by kibertoad** [comment]
@israelidanny Any chance you could still create the failing test? Sorry if you've moved on after all this time :)

**Comment by israelidanny** [comment]
@kibertoad yeah, I can! Still using knex with my little hack.
I will submit something next week.

**Comment by mercmobily** [comment]
Hello! Thanks for reporting this issue about question mark escaping in `knex.raw()`.

This issue has been thoroughly investigated and is tracked as a duplicate of the meta issue #6236, which covers the comprehensive problem with escaping of `?` in the codebase. I can see that this conversation evolved from the original complaint about `CONCAT(??)` with array bindings into a broader discussion about when question marks should and shouldn't be replaced in raw queries, particularly when no bindings are passed.


The core issue that @etodanik raised - that `knex.raw('')` without passing bindings should not replace question marks - is a valid concern that's part of the larger escaping problem. The investigation documented in #6236 shows this affects several scenarios:

1. When using named bindings with question marks in SQL string literals
2. JSONB path queries with `?` operators inside strings
3. Raw SQL scripts that contain question marks in regex patterns, comments, or string literals
4. Any case where a literal `?` appears in the SQL but no corresponding binding was intended


The current regex-based approach in `/lib/formatter/rawFormatter.js` cannot distinguish between question marks inside SQL string boundaries versus actual binding placeholders. Even though the `\?` escape sequence works for many cases, it fails when question marks appear in string literals or when using named bindings alongside literal question marks.


@kibertoad's suggestion that passing no bindings should mean no replacement is sensible and would help users who want to run SQL scripts through `knex.raw()` without needing to escape every question mark. However, this is just one aspect of the broader escaping problem.


Since this is covered by the meta issue #6236, this issue should be closed as a duplicate. All the context from this discussion (particularly the use case of running SQL scripts without bindings) has been considered in the comprehensive investigation. The meta issue tracks all related problems and potential solutions, including proper SQL-aware parsing to handle string boundaries correctly.


Related issues:
- #6236 (META issue - master)
- #3112 (child issue about `\?` producing syntax errors)
- #6163 (child issue about named bindings with `?` in queries)
- #519 (CLOSED - PostgreSQL JSONB operators)
- #2622 (CLOSED - question marks in values)
- #6190 (CLOSED - JSONB syntax)

**Comment by mercmobily** [comment]
Closing as duplicate of #6236
